src/main/java/org/truffleruby/core/hash/library/CompactHashStore.java

```diff
@@ -0,0 +1,506 @@
+    // occupied with valid data before.
+    private static final int INDEX_SLOT_UNUSED_MARKER = 0;
+    private static final int INDEX_SLOT_DELETED_MARKER = -1;
+
+    // Number of hash entries, not array positions (in general, capacities and sizes are always in entries)
+    public static final int DEFAULT_INITIAL_CAPACITY = 8;
+    public static final float THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD = 0.7f;
+
+    // temporary debugging aid so that its obvious the class is being used without being spam
+    private static boolean LOG_FIRST_ACCESS = false;
+
+    public CompactHashStore() {
+        this(DEFAULT_INITIAL_CAPACITY);
+    }
+
+    // private non-allocating constructor for .copy(), not part of the interface
+    private CompactHashStore(int[] index, Object[] kvs, int kvStoreInsertionPos, int numSlotsForIndexRebuild) {
+        this.index = index;
+        this.kvStore = kvs;
+        this.kvStoreInsertionPos = kvStoreInsertionPos;
+        this.numSlotsForIndexRebuild = numSlotsForIndexRebuild;
+    }
+
+    public CompactHashStore(int capacity) {
+        // temporary and ugly
+        if (LOG_FIRST_ACCESS) {
+            System.out.println("~~~~~~~~~~~~~~~~~~~~~~~ Using Compact Hashes from Hash Tables That Don't Hate You.");
+            LOG_FIRST_ACCESS = false; // without this we would spam the output stream
+        }
+        assertConstructorPreconditions(capacity);
+
+        int kvCapacity = roundUpwardsToNearestPowerOf2(capacity);
+        // the index array needs to be a little bit sparse for good performance (i.e. low load factor)
+        // so to store 8 key-value entries an index of exactly 8 slots is too crowded
+        int indexCapacity = 2 * kvCapacity;
+
+        // All 0s by default, so all slots are marked empty for free
+        this.index = new int[2 * indexCapacity];
+        this.kvStore = new Object[2 * kvCapacity];
+        this.kvStoreInsertionPos = 0; // always increases by 2, never decreases
+        this.numSlotsForIndexRebuild = (int) (this.index.length / 2 * THRESHOLD_LOAD_FACTOR_FOR_INDEX_REBUILD);
+    }
+
+    private static int roundUpwardsToNearestPowerOf2(int num) {
+        return Integer.highestOneBit(num - 1) << 1;
+    }
+
+    private static void assertConstructorPreconditions(int capacity) {
+        if (capacity < 1 || capacity > (1 << 28)) {
+            throw CompilerDirectives.shouldNotReachHere();
+        }
+    }
+
+    @ExportMessage
+    protected boolean set(RubyHash hash, Object key, Object value, boolean byIdentity,
+            @Cached @Shared HashingNodes.ToHash hashFunction,
+            @Cached @Shared GetHashPosInIndexNode getHashPos,
+            @Cached @Shared GetKeyPosInKVStoreNode getKey,
+            @Cached FreezeHashKeyIfNeededNode freezeKey,
+            @Cached @Exclusive InlinedConditionProfile keyNotFound,
+            @Bind("$node") Node __) { /* boilerplate argument for InlinedConditionProfile */
```

Comment:
Please stick to the convention used elsewhere in the runtime. Principally, it'd be `@Bind("$node") Node node`. The `__` isn't used anywhere else and makes it a bit hard to follow.

Reply:
Fair point, done.

I named it like that because I hated how it obscured the actual comparison in profiled If conditions, I wanted something with no letters to make the comparison stand out more and _ is already reserved.